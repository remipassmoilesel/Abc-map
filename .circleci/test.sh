#!/usr/bin/env bash

set -x
set -e

# TODO: add tests

mkdir -p /data/db && /usr/bin/mongod &

cd shared && npm install && npm run build

cd ..

cd api && npm install && npm run build

cd ..

cd gui && npm install && npm run build

