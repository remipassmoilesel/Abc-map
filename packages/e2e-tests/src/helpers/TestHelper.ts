/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Chainable = Cypress.Chainable;
import { Routes } from './Routes';

export class TestHelper {
  public static init(): Chainable<any> {
    return (
      cy
        // We set viewport
        .viewport(1980, 1080)
        // We set default language
        .visit(Routes.map().format(), {
          onBeforeLoad(win) {
            Object.defineProperty(win.navigator, 'language', { value: 'en-US' });
            Object.defineProperty(win.navigator, 'languages', { value: ['en'] });
            Object.defineProperty(win.navigator, 'accept_languages', { value: ['en'] });
          },
        })
        // We clear local storage data (and redux store persistence)
        // Despite docs says it run automatically before each test, it does not clear side menu states
        .clearLocalStorage()
    );
  }
}
