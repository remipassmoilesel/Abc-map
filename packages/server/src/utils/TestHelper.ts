/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import {
  AbcArtefact,
  AbcProjectManifest,
  AbcUser,
  ArtefactManifest,
  ArtefactType,
  CompressedProject,
  DEFAULT_PROJECTION,
  Language,
  LayerType,
  PredefinedLayerModel,
  ProjectConstants,
  Zipper,
} from '@abc-map/shared';
import * as uuid from 'uuid-random';
import { ProjectDocument } from '../projects/ProjectDocument';
import { DateTime } from 'luxon';
import { ArtefactManifestWithPath } from '../data-store/ArtefactManifestSchema';

export class TestHelper {
  public static sampleUser(): AbcUser {
    return {
      id: uuid(),
      email: `user-${uuid()}@test.ts`,
      password: 'what is a wr0ng passW0rd ????',
    };
  }

  public static sampleProject(): AbcProjectManifest {
    return {
      metadata: {
        id: uuid(),
        version: ProjectConstants.CurrentVersion,
        name: `Test project ${uuid()}`,
        public: false,
      },
      layers: [
        {
          type: LayerType.Predefined,
          metadata: {
            id: uuid(),
            name: 'OpenStreetMap',
            type: LayerType.Predefined,
            visible: true,
            active: false,
            opacity: 1,
            model: PredefinedLayerModel.OSM,
          },
        },
        {
          type: LayerType.Vector,
          metadata: {
            id: uuid(),
            name: 'Vecteurs',
            type: LayerType.Vector,
            visible: true,
            active: true,
            opacity: 1,
          },
          features: {
            type: 'FeatureCollection',
            features: [
              {
                id: uuid(),
                bbox: [1, 2, 3, 4],
                type: 'Feature',
                properties: {
                  val: 'var',
                },
                geometry: {
                  type: 'Point',
                  coordinates: [1, 5],
                },
              },
            ],
          },
        },
      ],
      layouts: {
        abcMapAttributionsEnabled: true,
        list: [],
      },
      view: {
        center: [1, 2],
        projection: DEFAULT_PROJECTION,
        resolution: 1000,
        rotation: 0,
      },
      sharedViews: {
        list: [],
        fullscreen: false,
        mapDimensions: {
          width: 800,
          height: 600,
        },
      },
    };
  }

  public static async sampleCompressedProject(): Promise<CompressedProject<Buffer>> {
    const project = this.sampleProject();
    const metadata = project.metadata;
    const zip = await Zipper.forNodeJS().zipFiles([{ path: ProjectConstants.ManifestName, content: Buffer.from(JSON.stringify(project), 'utf-8') }]);
    return {
      metadata: metadata,
      project: zip,
    };
  }

  public static sampleProjectDocument(): ProjectDocument {
    return {
      _id: uuid(),
      ownerId: uuid(),
      name: 'Fake project',
      version: '0.1',
      public: false,
    };
  }

  public static sampleArtefact(): AbcArtefact {
    return {
      id: uuid(),
      name: [{ language: Language.French, text: 'Burgers around the world' }],
      type: ArtefactType.Vector,
      license: 'AGPLv3.txt',
      attributions: ['Copyright somebody somewhere inc'],
      description: [{ language: Language.English, text: 'A beautiful artefact' }],
      files: ['/datastore/artefact/file1.gpx', '/datastore/artefact/file2.shp'],
      path: '/datastore/artefact/manifest.yml',
      provider: 'Nowhere inc',
      link: 'http://nowhere.net',
      previews: ['preview1.png'],
      keywords: [{ language: Language.English, text: ['beautiful', 'artifact'] }],
    };
  }

  public static sampleArtefactManifestWithPath(name: string): ArtefactManifestWithPath {
    return {
      version: '0.0.1',
      path: `/datastore/${name}/manifest.yml`,
      artefact: {
        name: [{ language: Language.English, text: name }],
        type: ArtefactType.BaseMap,
        license: `${name}-AGPLv3.txt`,
        attributions: ['Copyright somebody somewhere inc'],
        description: [{ language: Language.English, text: `A beautiful artefact named ${name}` }],
        keywords: [{ language: Language.English, text: ['beautiful', 'artifact', name] }],
        files: [`file1.gpx`, `file2.shp`],
        provider: `Nowhere inc`,
        link: `http://nowhere.net/${name}`,
      },
    };
  }

  public static sampleArtefactManifest(name: string): ArtefactManifest {
    return {
      version: '0.0.1',
      artefact: {
        name: [{ language: Language.English, text: name }],
        type: ArtefactType.BaseMap,
        license: `${name}-AGPLv3.txt`,
        attributions: ['Copyright somebody somewhere inc'],
        description: [{ language: Language.English, text: `A beautiful artefact named ${name}` }],
        keywords: [{ language: Language.English, text: ['beautiful', 'artifact', name] }],
        files: [`file1.gpx`, `file2.shp`],
        provider: `Nowhere inc`,
        link: `http://nowhere.net/${name}`,
      },
    };
  }

  public static randomDate(start: Date = new Date(1980, 1, 1, 0, 0, 0), end: Date = new Date(2020, 1, 1, 0, 0, 0)): DateTime {
    const timestamp = start.getTime() + Math.random() * (end.getTime() - start.getTime());
    return DateTime.fromMillis(timestamp);
  }

  /**
   * Warning: use this is a very bad idea
   */
  public static wait(timeMs: number): Promise<void> {
    return new Promise((resolve) => {
      setTimeout(resolve, timeMs);
    });
  }
}
