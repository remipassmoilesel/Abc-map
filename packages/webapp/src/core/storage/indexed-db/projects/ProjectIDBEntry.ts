/**
 * Copyright © 2022 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcProjectMetadata, AbcView } from '@abc-map/shared';

export const CURRENT_VERSION = 3;

export interface ProjectIDBEntry {
  version: number;
  // Metadata: id, version, etc ...
  metadata: AbcProjectMetadata;
  // Main view for edition
  view: AbcView;
  layerIds: string[];
  // Layouts for PDF exports
  layouts: {
    layoutIds: string[];
    abcMapAttributionsEnabled: boolean;
  };
  // Views publicly accessible
  sharedViews: {
    // If true, map is fullscreen
    fullscreen: boolean;
    // If not fullscreen, these dimensions are used
    mapDimensions: { width: number; height: number };
    // List of views
    viewIds: string[];
  };
}
