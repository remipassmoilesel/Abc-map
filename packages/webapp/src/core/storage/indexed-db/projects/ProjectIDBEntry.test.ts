/**
 * Copyright © 2022 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { TestHelper } from '../../../utils/test/TestHelper';
import { ProjectIDBEntry } from './ProjectIDBEntry';

describe('ProjectIDBEntry', () => {
  it('should not change without migration', () => {
    const template = TestHelper.sampleProjectManifest();
    const entry1: ProjectIDBEntry = {
      metadata: {
        ...template.metadata,
        id: 'test-project-id',
        name: 'Test project',
      },
      layouts: {
        layoutIds: ['layout-id-1', 'layout-id-2'],
        abcMapAttributionsEnabled: true,
      },
      sharedViews: {
        fullscreen: true,
        mapDimensions: {
          width: 888,
          height: 666,
        },
        viewIds: ['view-id-1', 'view-id-2'],
      },
      view: TestHelper.sampleView(),
      version: 2,
      layerIds: ['layer-id-1', 'layer-id-2'],
    };

    expect(entry1).toMatchSnapshot();
  });
});
