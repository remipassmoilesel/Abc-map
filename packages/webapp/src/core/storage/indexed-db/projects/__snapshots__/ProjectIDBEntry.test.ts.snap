// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectIDBEntry should not change without migration 1`] = `
Object {
  "layerIds": Array [
    "layer-id-1",
    "layer-id-2",
  ],
  "layouts": Object {
    "abcMapAttributionsEnabled": true,
    "layoutIds": Array [
      "layout-id-1",
      "layout-id-2",
    ],
  },
  "metadata": Object {
    "id": "test-project-id",
    "name": "Test project",
    "public": false,
    "version": "1.3.0",
  },
  "sharedViews": Object {
    "fullscreen": true,
    "mapDimensions": Object {
      "height": 666,
      "width": 888,
    },
    "viewIds": Array [
      "view-id-1",
      "view-id-2",
    ],
  },
  "version": 2,
  "view": Object {
    "center": Array [
      1.5,
      45.4,
    ],
    "projection": Object {
      "name": "EPSG:3857",
    },
    "resolution": 1000,
    "rotation": 0,
  },
}
`;
